File: scripts/lint/fix-imports.py
#!/usr/bin/env python3
"""
======= Automated Import Fixer =======

Reads the JSON output from check-cross-module-refs.py and automatically
fixes all resolvable issues by adding missing imports.

For each issue:
  - Finds which file exports the needed symbol
  - Adds the import to the file that needs it
  - Merges with existing imports from the same module when possible

Usage:
    python scripts/lint/check-cross-module-refs.py --json > issues.json
    python scripts/lint/fix-imports.py issues.json [--dry-run] [--verbose]

Exit code: 0 if all fixable issues resolved, 1 if some couldn't be fixed.
"""

import os
import re
import sys
import json
import argparse
from pathlib import Path
from collections import defaultdict
from typing import Optional

SCRIPT_DIR = Path(__file__).resolve().parent
PROJECT_ROOT = SCRIPT_DIR.parent.parent
JS_ROOT = PROJECT_ROOT / "public" / "assets" / "js"

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def resolve_import_path(from_rel: str, import_spec: str) -> str:
    """Resolve a relative import path to canonical form."""
    from_dir = os.path.dirname(from_rel)
    resolved = os.path.normpath(os.path.join(from_dir, import_spec))
    if not resolved.endswith('.js'):
        resolved += '.js'
    return resolved.replace('\\', '/')


def compute_import_spec(from_rel: str, to_rel: str) -> str:
    """Compute the relative import specifier from from_rel to to_rel."""
    from_dir = os.path.dirname(from_rel)
    rel = os.path.relpath(to_rel, from_dir).replace('\\', '/')
    if not rel.startswith('.'):
        rel = './' + rel
    # Remove .js extension for the import specifier
    if rel.endswith('.js'):
        rel = rel[:-3]
    return rel


def read_file_lines(rel_path: str) -> list[str]:
    """Read a file and return its lines."""
    abs_path = JS_ROOT / rel_path
    try:
        return abs_path.read_text(encoding='utf-8').split('\n')
    except Exception:
        return []


def write_file_lines(rel_path: str, lines: list[str]):
    """Write lines back to a file."""
    abs_path = JS_ROOT / rel_path
    abs_path.write_text('\n'.join(lines), encoding='utf-8')


def find_import_line(lines: list[str], module_spec: str) -> Optional[int]:
    """Find the line number (0-based) of an existing import from module_spec.
    Returns None if no existing import from that module."""
    # Normalize the module spec for comparison
    spec_pattern = module_spec.replace('.', r'\.')
    for i, line in enumerate(lines):
        stripped = line.strip()
        if stripped.startswith('import ') and 'from' in stripped:
            # Extract the module path
            m = re.search(r'from\s+[\'"]([^\'"]+)[\'"]', stripped)
            if m:
                existing_spec = m.group(1)
                # Compare resolved paths
                if existing_spec == module_spec:
                    return i
    return None


def add_import_to_line(line: str, symbol: str) -> str:
    """Add a symbol to an existing import statement's named imports.
    E.g., 'import { a, b } from "..."' -> 'import { a, b, symbol } from "..."'
    """
    # Find the opening brace
    m = re.search(r'import\s*\{([^}]*)\}\s*from', line)
    if not m:
        return line  # Can't handle non-named imports

    existing_names = m.group(1)
    # Parse existing names
    names = [n.strip() for n in existing_names.split(',') if n.strip()]
    if symbol in names:
        return line  # Already imported

    # Check if we need to add a line break (long imports)
    if len(','.join(names)) > 60:
        # Multi-line format
        new_import = line.replace(
            '{' + existing_names + '}',
            '{\n' + ',\n'.join('  ' + n for n in names) + ',\n  ' + symbol + '\n}'
        )
    else:
        new_import = line.replace(
            '{' + existing_names + '}',
            '{' + ', '.join(names + [symbol]) + '}'
        )
    return new_import


def create_import_line(symbol: str, module_spec: str) -> str:
    """Create a new import statement line."""
    return f'import {{ {symbol} }} from "{module_spec}";'


def find_insert_position(lines: list[str]) -> int:
    """Find the best position to insert a new import line (after last import)."""
    last_import = -1
    for i, line in enumerate(lines):
        stripped = line.strip()
        if stripped.startswith('import ') and 'from' in stripped:
            last_import = i
        elif stripped.startswith('import ') and 'from' not in stripped:
            last_import = i
    return last_import + 1 if last_import >= 0 else 0


# ---------------------------------------------------------------------------
# Cross-reference builder (reuses logic from check-cross-module-refs.py)
# ---------------------------------------------------------------------------

def build_exports_map(js_root: Path) -> dict[str, set[str]]:
    """Build a map of file -> set of exported symbol names.
    Parses export statements from all .js files under js_root."""
    exports: dict[str, set[str]] = defaultdict(set)

    for root, dirs, files in os.walk(js_root):
        dirs[:] = [d for d in dirs if not d.startswith('.')]
        for fname in files:
            if not fname.endswith('.js'):
                continue
            filepath = Path(root) / fname
            try:
                content = filepath.read_text(encoding='utf-8')
            except Exception:
                continue
            rel_path = str(filepath.relative_to(js_root)).replace('\\', '/')

            # Parse exports
            for m in re.finditer(
                r'export\s+(?:(?:async\s+)?function\s+(\w+)|'
                r'(?:const|let|var)\s+(\w+)|'
                r'class\s+(\w+))',
                content
            ):
                name = m.group(1) or m.group(2) or m.group(3)
                if name:
                    exports[rel_path].add(name)

            # export { a, b }
            for m in re.finditer(r'export\s*\{([^}]*)\}', content):
                for name in re.findall(r'\b(\w+)\b', m.group(1)):
                    exports[rel_path].add(name)

    return dict(exports)


# ---------------------------------------------------------------------------
# Main fixer logic
# ---------------------------------------------------------------------------

def fix_issues(issues: list[dict], exports_map: dict[str, set[str]], dry_run: bool = False, verbose: bool = False) -> dict:
    """Fix all resolvable issues. Returns stats."""
    stats = {'fixed': 0, 'skipped_no_source': 0, 'skipped_already_imported': 0, 'errors': 0}

    # Filter to missing-import issues only (most actionable)
    missing = [i for i in issues if i['type'] == 'missing-import']

    # Also handle undefined-reference issues that have suggestions
    undef_with_suggestions = [
        i for i in issues
        if i['type'] == 'undefined-reference' and i.get('suggestion')
    ]

    all_fixable = missing + undef_with_suggestions

    # Group fixes by (file, module_to_import_from) to batch them
    # This way we add multiple symbols to one import line
    fixes_by_file: dict[str, dict[str, set[str]]] = defaultdict(lambda: defaultdict(set))

    for issue in all_fixable:
        file = issue['file']
        symbol = issue['symbol']

        # For missing imports: find source
        if issue['type'] == 'missing-import':
            # Find which file exports this symbol
            sources = [f for f, exps in exports_map.items() if symbol in exps and f != file]
            if not sources:
                stats['skipped_no_source'] += 1
                if verbose:
                    print(f"  ⚠ {file}:{issue['line']} — '{symbol}' not exported by any file")
                continue
            # Pick the best (shortest relative path) source
            best_source = min(sources, key=lambda s: len(
                os.path.relpath(s, os.path.dirname(file)).replace('\\', '/')
            ))
            import_spec = compute_import_spec(file, best_source)
            fixes_by_file[file][import_spec].add(symbol)

        # For undefined references with suggestion
        elif issue['type'] == 'undefined-reference' and issue.get('suggestion'):
            suggestion = issue['suggestion']
            # The suggestion is already a relative import path like './foo.js'
            # Remove .js extension for the import specifier
            if suggestion.endswith('.js'):
                suggestion = suggestion[:-3]
            fixes_by_file[file][suggestion].add(symbol)

    # Apply fixes
    for file, module_imports in sorted(fixes_by_file.items()):
        lines = read_file_lines(file)
        if not lines:
            if verbose:
                print(f"  ❌ {file}: could not read file")
            stats['errors'] += 1
            continue

        modified = False
        for module_spec, symbols in sorted(module_imports.items()):
            existing_line = find_import_line(lines, module_spec)

            if existing_line is not None:
                # Add to existing import
                old_line = lines[existing_line]
                new_line = old_line
                for sym in sorted(symbols):
                    new_line = add_import_to_line(new_line, sym)
                if new_line != old_line:
                    lines[existing_line] = new_line
                    modified = True
                    if verbose:
                        print(f"  ✅ {file}: added {symbols} to existing import from '{module_spec}'")
            else:
                # Create new import line
                insert_pos = find_insert_position(lines)
                for sym in sorted(symbols):
                    new_import = create_import_line(sym, module_spec)
                    lines.insert(insert_pos, new_import)
                    insert_pos += 1
                    modified = True
                    if verbose:
                        print(f"  ✅ {file}: added import {{ {sym} }} from '{module_spec}'")

        if modified:
            stats['fixed'] += len(symbols) if 'symbols' in dir() else 1
            if not dry_run:
                write_file_lines(file, lines)
            else:
                if verbose:
                    print(f"  📝 {file}: would write changes (dry-run)")

    return stats


def main():
    parser = argparse.ArgumentParser(description="Automated import fixer")
    parser.add_argument('issues_json', help='Path to JSON issues file from check-cross-module-refs.py')
    parser.add_argument('--dry-run', action='store_true', help='Show what would be changed without applying')
    parser.add_argument('--verbose', '-v', action='store_true', help='Show detailed fix information')
    args = parser.parse_args()

    # Load issues
    try:
        issues = json.loads(Path(args.issues_json).read_text(encoding='utf-8'))
    except Exception as e:
        print(f"❌ Could not read issues file: {e}", file=sys.stderr)
        sys.exit(2)

    print(f"📋 Loaded {len(issues)} issues from {args.issues_json}")

    # Build exports map
    print(f"🔍 Building exports map from {JS_ROOT}...")
    exports_map = build_exports_map(JS_ROOT)
    print(f"   Found {len(exports_map)} files with exports")

    if args.dry_run:
        print("\n⚠ DRY RUN — no files will be modified\n")

    # Fix issues
    stats = fix_issues(issues, exports_map, dry_run=args.dry_run, verbose=args.verbose)

    # Summary
    print(f"\n{'='*60}")
    print(f"📊 FIX SUMMARY:")
    print(f"   Fixed:               {stats['fixed']}")
    print(f"   Skipped (no source): {stats['skipped_no_source']}")
    print(f"   Errors:              {stats['errors']}")
    print(f"{'='*60}")

    if stats['skipped_no_source'] > 0:
        print("\n⚠ Some symbols are not exported from any file.")
        print("  These need 'export' added to their source files first.")
        print("  Then re-run this script.")

    sys.exit(0 if stats['errors'] == 0 else 1)


if __name__ == '__main__':
    main()

File: scripts/lint/fix-imports.js
/**
 * ======= Auto-Import Fixer =======
 *
 * Scans all exports across the JS codebase, then automatically adds
 * missing imports to every file that calls functions without importing them.
 *
 * Usage: node scripts/lint/fix-imports.js [--dry-run] [--verbose]
 */

const fs = require("fs");
const path = require("path");

const JS_ROOT = path.resolve(__dirname, "..", "..", "public", "assets", "js");
const dryRun = process.argv.includes("--dry-run");
const verbose = process.argv.includes("--verbose");

// ---------------------------------------------------------------
// Step 1: Build export map — { functionName: relativePath }
// ---------------------------------------------------------------

const RE_EXPORT_FN = /export\s+(?:async\s+)?function\s+(\w+)/g;
const RE_EXPORT_CONST = /export\s+(?:const|let|var)\s+(\w+)/g;
const RE_EXPORT_CLASS = /export\s+class\s+(\w+)/g;

function buildExportMap() {
  const map = new Map(); // name → [file1, file2, ...] (rarely multiple)

  function* walk(dir) {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        yield* walk(full);
      } else if (entry.name.endsWith(".js")) {
        yield full;
      }
    }
  }

  for (const filePath of walk(JS_ROOT)) {
    const rel = path.relative(JS_ROOT, filePath).replace(/\\/g, "/");
    if (rel === "data/state.js") continue; // skip

    const source = fs.readFileSync(filePath, "utf-8");

    for (const m of source.matchAll(RE_EXPORT_FN)) {
      const name = m[1];
      if (!map.has(name)) map.set(name, []);
      if (!map.get(name).includes(rel)) map.get(name).push(rel);
    }
    for (const m of source.matchAll(RE_EXPORT_CONST)) {
      const name = m[1];
      if (!map.has(name)) map.set(name, []);
      if (!map.get(name).includes(rel)) map.get(name).push(rel);
    }
    for (const m of source.matchAll(RE_EXPORT_CLASS)) {
      const name = m[1];
      if (!map.has(name)) map.set(name, []);
      if (!map.get(name).includes(rel)) map.get(name).push(rel);
    }
  }

  return map;
}

// ---------------------------------------------------------------
// Step 2: Get relative import path from file A to file B
// ---------------------------------------------------------------
function getRelativeImportPath(fromFileRel, toFileRel) {
  const fromDir = path.dirname(fromFileRel);
  let rel = path.relative(fromDir, toFileRel).replace(/\\/g, "/");
  if (!rel.startsWith(".")) rel = "./" + rel;
  return rel.replace(/\.js$/, ".js"); // keep .js extension
}

// ---------------------------------------------------------------
// Step 3: Find what's missing in a file and add imports
// ---------------------------------------------------------------

function fixFile(filePath, exportMap) {
  const relPath = path.relative(JS_ROOT, filePath).replace(/\\/g, "/");
  const source = fs.readFileSync(filePath, "utf-8");

  // Skip if no function calls at all
  if (source.length < 10) return { changed: false, path: relPath };

  // Extract existing imports
  const existingImports = new Set();
  const RE_NAMED_IMPORT = /import\s+\{([^}]+)\}\s*from\s*['"]([^'"]+)['"]/g;
  for (const m of source.matchAll(RE_NAMED_IMPORT)) {
    const names = m[1].split(",").map((s) => {
      const parts = s.trim().split(/\s+as\s+/);
      return parts[parts.length - 1].trim();
    });
    for (const n of names) existingImports.add(n);
  }

  // Extract local definitions (function/const/let/var)
  const localDefs = new Set();
  const RE_FN = /(?:^|\s)(?:async\s+)?function\s+(\w+)/gm;
  const RE_VAR = /(?:^|\s)(?:const|let|var)\s+(\w+)\s*=/gm;
  const RE_CLASS = /(?:^|\s)class\s+(\w+)/gm;
  for (const m of source.matchAll(RE_FN)) localDefs.add(m[1]);
  for (const m of source.matchAll(RE_VAR)) localDefs.add(m[1]);
  for (const m of source.matchAll(RE_CLASS)) localDefs.add(m[1]);

  // Extract function calls (simple identifiers followed by paren)
  // Strip strings and comments first
  let clean = source.replace(/`[^`]*`/g, '""');
  clean = clean.replace(/"(?:[^"\\]|\\.)*"/g, '""');
  clean = clean.replace(/'(?:[^'\\]|\\.)*'/g, '""');
  clean = clean.replace(/\/\/.*$/gm, "");
  clean = clean.replace(/\/\*[\s\S]*?\*\//g, "");

  const RE_CALL = /(?<![.\w])(\w+)\s*\(/g;
  const calledNames = new Set();
  for (const m of clean.matchAll(RE_CALL)) {
    const name = m[1];
    if (/^(if|for|while|switch|catch|function|return|typeof|new|throw|delete|void|await|yield|case|import|export|class|extends|super|instanceof|in|of|try|finally|else|do|break|continue|debugger|with|static|async|get|set)$/.test(name)) continue;
    calledNames.add(name);
  }

  // Known globals that we should NOT import
  const KNOWN_GLOBALS = new Set([
    "Object", "Array", "String", "Number", "Boolean", "Date", "Math",
    "JSON", "Error", "TypeError", "RegExp", "Map", "Set", "WeakMap",
    "Promise", "Symbol", "Intl", "console", "document", "window",
    "navigator", "localStorage", "sessionStorage", "location", "fetch",
    "setTimeout", "setInterval", "clearTimeout", "clearInterval",
    "requestAnimationFrame", "cancelAnimationFrame",
    "addEventListener", "removeEventListener",
    "parseInt", "parseFloat", "isNaN", "isFinite", "eval",
    "encodeURIComponent", "decodeURIComponent",
    "FormData", "FileReader", "Blob", "URL", "URLSearchParams",
    "IntersectionObserver", "MutationObserver", "ResizeObserver",
    "CustomEvent", "Event",
    "firebase", "$", "jQuery", "bootstrap", "AOS", "Swiper", "GLightbox",
    "Isotope", "Typed", "Waypoint", "Iconify", "google", "Sortable",
    "XMLHttpRequest", "ClipboardItem", "Image",
    "require", "module", "exports", "__dirname", "__filename", "process",
    "global", "globalThis", "alert", "confirm", "prompt",
    "gsap", "ScrollTrigger",
    // Callback/parameter names
    "fn", "callback", "done", "resolve", "reject", "action", "check",
    "build", "batch", "task", "constructor",
  ]);

  // Find missing: called but not defined locally and not imported
  const missing = [];
  for (const name of calledNames) {
    if (localDefs.has(name)) continue;
    if (existingImports.has(name)) continue;
    if (KNOWN_GLOBALS.has(name)) continue;
    if (!exportMap.has(name)) continue; // can't resolve

    const sources = exportMap.get(name);
    // Pick the first source (most common)
    const sourceFile = sources[0];
    if (sourceFile === relPath) continue; // same file

    missing.push({ name, sourceFile });
  }

  if (missing.length === 0) return { changed: false, path: relPath };

  // Group missing by source file
  const bySource = new Map();
  for (const m of missing) {
    if (!bySource.has(m.sourceFile)) bySource.set(m.sourceFile, []);
    bySource.get(m.sourceFile).push(m.name);
  }

  // Deduplicate names
  for (const [src, names] of bySource) {
    bySource.set(src, [...new Set(names)]);
  }

  // Build new import lines and merge with existing
  const newImportLines = [];
  const mergeIntoExisting = []; // { importPath, lineIndex, namesToAdd }

  for (const [sourceFile, names] of bySource) {
    const importPath = getRelativeImportPath(relPath, sourceFile);
    const sortedNames = names.sort();

    // Check if we already import from this path
    const existingLineRegex = new RegExp(
      `import\\s+\\{([^}]+)\\}\\s*from\\s*['"]${importPath.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}['"]`,
      "g"
    );
    const existingMatch = existingLineRegex.exec(source);

    if (existingMatch) {
      // Merge new names into existing import
      const existingNames = existingMatch[1]
        .split(",")
        .map((s) => s.trim().split(/\s+as\s+/).pop().trim())
        .filter(Boolean);

      const allNames = [...new Set([...existingNames, ...sortedNames])].sort();
      const newImportLine = `import { ${allNames.join(", ")} } from '${importPath}';`;

      // Find the line index in the source
      const beforeMatch = source.substring(0, existingMatch.index);
      const lineIndex = beforeMatch.split("\n").length - 1;
      mergeIntoExisting.push({
        lineIndex,
        oldLine: existingMatch[0],
        newLine: newImportLine,
      });
    } else {
      newImportLines.push(
        `import { ${sortedNames.join(", ")} } from '${importPath}';`
      );
    }
  }

  if (newImportLines.length === 0 && mergeIntoExisting.length === 0) {
    return { changed: false, path: relPath };
  }

  // Apply merges first
  let mergedSource = source;
  for (const merge of mergeIntoExisting) {
    mergedSource = mergedSource.replace(merge.oldLine, merge.newLine);
  }

  // Insert new imports
  const lines = mergedSource.split("\n");
  let lastImportIdx = 0;
  for (let i = 0; i < lines.length; i++) {
    if (/^\s*import\s/.test(lines[i])) {
      lastImportIdx = i + 1;
    }
  }

  // If no existing imports, put after the header comment block
  if (lastImportIdx === 0) {
    // Find end of comment block
    for (let i = 0; i < lines.length; i++) {
      if (lines[i].includes("*/")) {
        lastImportIdx = i + 1;
        break;
      }
    }
    if (lastImportIdx === 0) lastImportIdx = 0;
    newImportLines.push(""); // blank line after imports
  }

  lines.splice(lastImportIdx, 0, ...newImportLines);
  const newSource = lines.join("\n");

  if (!dryRun) {
    fs.writeFileSync(filePath, newSource, "utf-8");
  }

  const totalAdded = missing.length;
  return {
    changed: true,
    path: relPath,
    added: totalAdded,
    lines: [...newImportLines, ...mergeIntoExisting.map((m) => m.newLine)],
  };
}

// ---------------------------------------------------------------
// Main
// ---------------------------------------------------------------
console.log(`\n🔧 Auto-Import Fixer`);
console.log(`   Mode: ${dryRun ? "DRY RUN" : "LIVE"}\n`);

// Build export map
console.log("   Building export map...");
const exportMap = buildExportMap();
console.log(`   Found ${exportMap.size} exported names.\n`);

// Process all files
const results = [];
function* walk(dir) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      yield* walk(full);
    } else if (entry.name.endsWith(".js") && !entry.name.includes(".test.")) {
      yield full;
    }
  }
}

for (const filePath of walk(JS_ROOT)) {
  const result = fixFile(filePath, exportMap);
  if (result.changed) {
    results.push(result);
    if (verbose) {
      console.log(`  ✅ ${result.path} (+${result.added} imports)`);
      for (const line of result.lines) {
        console.log(`     ${line}`);
      }
    }
  }
}

console.log(`\n${"=".repeat(60)}`);
console.log(`   Files changed: ${results.length}`);
console.log(`   Total imports added: ${results.reduce((s, r) => s + r.added, 0)}`);
if (!verbose && results.length > 0) {
  for (const r of results) {
    console.log(`   ✅ ${r.path} (+${r.added})`);
  }
}
console.log(`${"=".repeat(60)}\n`);

if (dryRun) {
  console.log("⚠️  DRY RUN — no files were modified.");
  console.log("   Run without --dry-run to apply changes.\n");
}
